packages/runtime/container-runtime/src/gc/gcTelemetry.ts

```diff
@@ -119,76 +124,111 @@ export class GCTelemetryTracker {
+		// If the node's unreferenced state is not tracked or it's not tombstoned, there is nothing to log.
+		const nodeStateTracker = this.getNodeStateTracker(nodeUsageProps.id);
+		if (nodeStateTracker === undefined && !nodeUsageProps.isTombstoned) {
 			return;
 		}
 
-		// Add the unique event id so that we don't generate a log for this event again in this session..
-		this.loggedUnreferencedEvents.add(uniqueEventId);
-
-		const state = nodeStateTracker.state;
+		const nodeType = this.getNodeType(nodeUsageProps.id);
 		const { usageType, currentReferenceTimestampMs, packagePath, id, fromId, ...propsToLog } =
 			nodeUsageProps;
-		const eventProps: Omit<IUnreferencedEventProps, "state" | "usageType"> = {
+		const tombstoneProps: Partial<IUnreferencedEventProps> = {
 			type: nodeType,
-			unrefTime: nodeStateTracker.unreferencedTimestampMs,
+			unrefTime: nodeStateTracker?.unreferencedTimestampMs,
 			age:
-				nodeUsageProps.currentReferenceTimestampMs -
-				nodeStateTracker.unreferencedTimestampMs,
+				nodeStateTracker !== undefined
+					? nodeUsageProps.currentReferenceTimestampMs -
+					  nodeStateTracker.unreferencedTimestampMs
+					: undefined,
 			timeout:
-				state === UnreferencedState.Inactive
+				nodeStateTracker?.state === UnreferencedState.Inactive
 					? this.configs.inactiveTimeoutMs
 					: this.configs.sweepTimeoutMs,
 			...tagCodeArtifacts({ id, fromId }),
 			...propsToLog,
 			...this.createContainerMetadata,
 		};
 
-		// This will log the following events:
-		// GC_Tombstone_DataStore_Revived, GC_Tombstone_SubDataStore_Revived, GC_Tombstone_Blob_Revived
-		if (nodeUsageProps.usageType === "Revived" && nodeUsageProps.isTombstoned) {
-			sendGCUnexpectedUsageEvent(
-				this.mc,
-				{
-					eventName: `GC_Tombstone_${nodeType}_Revived`,
-					category: "generic",
-					...tagCodeArtifacts({ url: id }),
-					gcTombstoneEnforcementAllowed: this.configs.tombstoneEnforcementAllowed,
-				},
-				undefined /* packagePath */,
-			);
+		// If the node is tombstoned, log telemetry. This will log the following events:
+		// GC_Tombstone_DataStore_Requested, GC_Tombstone_DataStore_Changed, GC_Tombstone_DataStore_Revived
+		// GC_Tombstone_SubDataStore_Requested, GC_Tombstone_SubDataStore_Changed, GC_Tombstone_SubDataStore_Revived
+		// GC_Tombstone_Blob_Requested, GC_Tombstone_Blob_Changed, GC_Tombstone_Blob_Revived
+		if (nodeUsageProps.isTombstoned) {
+			const { id: taggedId, fromId: taggedFromId, headers, ...otherProps } = tombstoneProps;
+			const eventUsageName = usageType === "Loaded" ? "Requested" : usageType;
+			const event = {
+				eventName: `GC_Tombstone_${nodeType}_${eventUsageName}`,
+				pkg: tagCodeArtifacts({ pkg: nodeUsageProps.packagePath?.join("/") }).pkg,
+				stack: generateStack(),
+				id: taggedId,
+				fromId: taggedFromId,
+				headers: JSON.stringify(headers),
+				details: JSON.stringify({
+					...otherProps,
+				}),
+				gcTombstoneEnforcementAllowed: this.configs.tombstoneEnforcementAllowed,
```

Comment:
You actually can ditch the JSON.stringify here.  It's a little awkward with `headers`, if you try putting it there as-is it complains because it doesn't have an index signature like the telemetry props thing does.  This works though:


```suggestion
				headers: { ...headers },
				details: otherProps,
				gcTombstoneEnforcementAllowed: this.configs.tombstoneEnforcementAllowed,
```